include/revng/ADT/GenericGraph.h

```diff
@@ -996,6 +996,124 @@ class MutableEdgeNode
 
+private:
+  /// Custom child iterator which iterates over the concatenation of the
+  /// successors and predecessors of a node. This is used for the `Undirected`
+  /// `GraphTraits` specialization. This is done taking inspiration from the
+  /// internals of `llvm::concat` implementation, but overcoming the limitation
+  /// causing by the return `EdgeView` going of out scope.
+  template<typename SuccessorIterator, typename PredecessorIterator>
+  class UndirectedChildIterator
+    : public llvm::iterator_facade_base<
+        UndirectedChildIterator<SuccessorIterator, PredecessorIterator>,
+        std::bidirectional_iterator_tag,
+        EdgeView> {
+  private:
+    SuccessorIterator SBegin, SEnd;
+    PredecessorIterator PBegin, PEnd;
+
+    explicit UndirectedChildIterator(SuccessorIterator SBegin,
+                                     SuccessorIterator SEnd,
+                                     PredecessorIterator PBegin,
+                                     PredecessorIterator PEnd) :
+      SBegin(SBegin), SEnd(SEnd), PBegin(PBegin), PEnd(PEnd) {}
+
+  public:
+    static UndirectedChildIterator<SuccessorIterator, PredecessorIterator>
+    begin(llvm::iterator_range<SuccessorIterator> Successors,
+          llvm::iterator_range<PredecessorIterator> Predecessors) {
+      return UndirectedChildIterator<
+        SuccessorIterator,
+        PredecessorIterator>(std::begin(Successors),
+                             std::end(Successors),
+                             std::begin(Predecessors),
+                             std::end(Predecessors));
+    }
+    static UndirectedChildIterator<SuccessorIterator, PredecessorIterator>
+    end(llvm::iterator_range<SuccessorIterator> Successors,
+        llvm::iterator_range<PredecessorIterator> Predecessors) {
+      return UndirectedChildIterator<
+        SuccessorIterator,
+        PredecessorIterator>(std::end(Successors),
+                             std::end(Successors),
+                             std::end(Predecessors),
+                             std::end(Predecessors));
+    }
+
+    using UndirectedChildIterator::iterator_facade_base::operator++;
+
+    UndirectedChildIterator &operator++() {
+      if (SBegin != SEnd) {
+        ++SBegin;
+        return *this;
+      }
+
+      if (PBegin != PEnd) {
+        ++PBegin;
+        return *this;
+      }
+
+      revng_abort("Attempted to increment an end iterator!");
+    }
+
+    decltype(auto) operator*() {
+      if (SBegin != SEnd)
+        return *SBegin;
+
+      if (PBegin != PEnd)
+        return *PBegin;
+
+      revng_abort("Attempted to dereference an end iterator!");
+    }
+    decltype(auto) operator*() const {
+      if (SBegin != SEnd)
+        return *SBegin;
+
+      if (PBegin != PEnd)
+        return *PBegin;
+
+      revng_abort("Attempted to dereference an end iterator!");
+    }
+
+    // Proxy object which extends the lifetime of the `EdgeView` object returned
+    // by the underlying `map_iterator`
```

Comment:
I'm not sure what `map_iterator` we're talking about here. This piece of code seems to assume some specific usage.

Reply:
Yeah, this was more confusing than anything, I centralized the explanation for having this at the beginning of the class, as stated in the comment above https://github.com/revng/revng/pull/388#discussion_r1705687147